component/metrics/receiver.go

```diff
@@ -13,9 +14,35 @@ type Receiver struct {
 // RiverCapsule marks receivers as a capsule.
 func (r Receiver) RiverCapsule() {}
 
-// FlowMetric is a wrapper around a single metric without the timestamp
+// FlowMetric is a wrapper around a single metric without the timestamp, FlowMetric is *mostly* immutable
+// the globalrefid can only be set if it is 0
 type FlowMetric struct {
-	GlobalRefID uint64
-	Labels      labels.Labels
-	Value       float64
+	globalRefID *atomic.Uint64
+	labels      labels.Labels
+	value       float64
+}
+
+// NewFlowMetric instantiates a new flow metric
+func NewFlowMetric(globalRefID uint64, lbls labels.Labels, value float64) *FlowMetric {
+	return &FlowMetric{
+		globalRefID: atomic.NewUint64(globalRefID),
+		labels:      lbls,
+		value:       value,
+	}
+}
+
+// GlobalRefID Retrieves the GlobalRefID
+func (fw *FlowMetric) GlobalRefID() uint64 { return fw.globalRefID.Load() }
+
+// SetGlobalRefID will only set the GlobalRefID if it is 0
+func (fw *FlowMetric) SetGlobalRefID(globalRefID uint64) {
+	fw.globalRefID.CAS(0, globalRefID)
```

Comment:
The CAS only works if the previous ref ID was 0, which it almost never will be.

Reply:
That is by design, if we don't want the double checking then we can make it immutable.

Reply:
Sorry, I missed the doc comment. The API here feels a little hard to follow, I think we should spend some time investing in making it easier to reason about.